### PR TITLE
Rtc

### PR DESCRIPTION
**Pull request by maikf**
Don't merge yet, please review :)

### CODE REVIEW

File: src/c128time.c [hunk 1]
```diff
@@ -7,33 +7,83 @@
...
+    }
+  } else {
+    hour = bcd2dec(bcd_hour);
+  }
+
+  sec = bcd2dec(CIA1.tod_sec);
+  min = bcd2dec(CIA1.tod_min);
+
+  /* MUST read tod_10 to enable the clock latch again */
+  tenth = CIA1.tod_10;
```
Comment (stapelberg) [inline src/c128time.c]: I suggest `(void)CIA1.tod_10;` to signal that the value is discarded immediately

File: src/c128time.c [hunk 2]
```diff
@@ -7,33 +7,83 @@
...
-                    ((uint32_t)hrs * (uint32_t)3600)) *
-                   (uint32_t)60;
-  uint32_t lowbit = (added & 0xFF);
-  uint32_t middlebit = (added >> 8) & 0xFF;
-  uint32_t highbit = (added >> 16) & 0xFF;
-
-  POKE(0x00A0, (BYTE)highbit);
-  POKE(0x00A1, (BYTE)middlebit);
-  POKE(0x00A2, (BYTE)lowbit);
+/* divide by 10; put quotient in high nibble, reminder in low nibble */
```
Comment (stapelberg) [inline src/c128time.c]: s/reminder/remainder/

File: src/c128time.c [hunk 3]
```diff
@@ -7,33 +7,83 @@
...
+uint8_t dec2bcd(uint8_t dec) { return (((dec / 10) << 4) | (dec % 10)); }
```
Comment (stapelberg) [inline src/c128time.c]: Make this `const uint8_t dec`

File: src/c128time.c [hunk 4]
```diff
@@ -7,33 +7,83 @@
...
+
+void set_time(uint8_t day, uint8_t hour, uint8_t min, uint8_t sec) {
+  uint8_t bcd_hour;
+
+  /* CIA TOD will always flip the pm bit
+   * when either 0 or 12 is written to the hour register */
+  if (hour == 0) {
+    /* bcd 12 with high bit (pm) set */
+    bcd_hour = 0x92;
```
Comment (stapelberg) [inline src/c128time.c]: `dec2bcd(12) & 0x80` seems easier to read to me

File: src/c128time.c [hunk 5]
```diff
@@ -7,33 +7,83 @@
...
+  } else if (hour > 12) {
+    /* convert 24h clock to 12h with pm bit set */
+    bcd_hour = dec2bcd(hour - 12);
+    bcd_hour = bcd_hour ^ 0x80;
+  } else {
+    /* includes 12pm since the bit gets automatically flipped */
```
Comment (stapelberg) [inline src/c128time.c]: Which bit gets flipped automatically?

File: src/kasse.c [hunk 6]
```diff
@@ -269,21 +269,24 @@ void buy_custom(void) {
...
 
 void set_time_interactive(void) {
-  BYTE part[3] = {'0', '0', '\0'};
-  BYTE tp1, tp2, tp3;
+  char part[3] = {'\0', '\0', '\0'};
+  uint8_t day, tp1, tp2, tp3;
   char *time_input, *time;
-  cprintf("Gib die aktuelle Uhrzeit ein (Format HHMMSS):\r\n");
+  cprintf("Gib den aktuellen Tag des Events und Uhrzeit ein\r\n"
+          "(Format DHHMMSS):\r\n");
```
Comment (stapelberg) [inline src/kasse.c]: One character for the day? Does the day start at 0 or 1? Can we add an example value?

File: src/c128time.c [hunk 7]
```diff
@@ -7,33 +7,83 @@
...
-  BYTE hrs, min;
+#include "bcd2dec.h"
+#include "general.h"
+#include "globals.h"
+
+void update_time(void) {
+  uint8_t bcd_hour, hour, min, sec, tenth;
+
+  /* Read the hour register first to stop the clock from updating the external
+   * registers from the internal (still ticking!) CIA registers. */
```
Comment (stapelberg) [inline src/c128time.c]: Here and elsewhere: what does CIA stand for? Can we expand the acronym?

Comment (9b17fe) [reply src/c128time.c]: The Complex Interface Adapter. Should I maybe link this? 
https://www.c64-wiki.com/wiki/CIA

File: src/kasse.c (hunk 6 above)
Comment (9b17fe) [reply src/kasse.c]: yeah, I haven't thought about this yet. What do you propose?

File: src/c128time.c (hunk 1 above)
Comment (9b17fe) [reply src/c128time.c]: This will be optimized away. I guess the saner alternative would be inline assembler.

File: src/c128time.c (hunk 7 above)
Comment (sur5r) [reply src/c128time.c]: Given we're on a Commodore machine it's pretty obvious, but the link does not hurt either way.

File: src/c128time.c (hunk 1 above)
Comment (sur5r) [reply src/c128time.c]: This would not happen if CIA1 were declared volatile. Should we open an issue at cc65/cc65?

File: src/c128time.c [hunk 8]
```diff
@@ -87,3 +102,25 @@ void set_time(uint8_t day, uint8_t hour, uint8_t min, uint8_t sec) {
...
+  static uint8_t x, y;
+  /* We are called 60 times a second. We only want to draw a clock
+   * when we are a) on the mainscreen and b) the seconds changed */
+  if (kasse_menu == MENU_MAIN && CIA1.tod_sec != daytime.sec) {
+    t = get_time();
+    x = wherex();
+    y = wherey();
+    cputsxy(70, 3, t);
+    gotoxy(x, y);
+  }
```
Comment (sur5r) [inline src/c128time.c]: We should definitely not write to the screen from an IRQ handler. That's way too much code, IMHO.

File: src/c128time.c [hunk 9]
```diff
@@ -9,23 +9,35 @@
...
+
+/* the Time of Day PM bit is set for hours >= 12 */
+#define TOD_PM 0x80
+
+#define DAYTIME_IRQ_STACK_SIZE 32
+uint8_t daytime_irq_stack[DAYTIME_IRQ_STACK_SIZE];
+
 void update_time(void) {
-  uint8_t bcd_hour, hour, min, sec, tenth;
+  volatile static uint8_t bcd_hour, hour, min, sec, tenth;
```
Comment (sur5r) [inline src/c128time.c]: This is unneeded. There's no concurrent access to this data.

File: src/c128time.c (hunk 8 above)
Comment (9b17fe) [reply src/c128time.c]: I don't think it is too big a deal:
- The limiting factor is not the amount of code, but the slow interface to the VDC.
- calling 1000 iterations of current code (I removed the check if seconds changed) take about 10 seconds, that's tight but still within 16ms per IRQ handler.
- directly using vdc_write_mem() takes this down to 7 seconds for 1000 interations or 7ms per IRQ handler.

There's nothing happening in the code outside the IRQ handler, it just sits there at cgetc();

File: src/c128time.c (hunk 7 above)
Comment (stapelberg) [reply src/c128time.c]: Yeah, adding the link would be great :)

File: src/c128time.c (hunk 1 above)
Comment (stapelberg) [reply src/c128time.c]: Yeah, we should.

File: src/kasse.c (hunk 6 above)
Comment (stapelberg) [reply src/kasse.c]: I’d suggest to start at day 0. How about:
```
(Format DHHMMSS, 0-indexiert, z.B. 0174259):\r\n
```

**Pull request merged**